GCS_MAVLink: don't lie about bad AHRS

IamPete1 opened the pull request:
This is a logic change to how the AHRS sys status bit is set. To report healthy it must now be healthy and initialized. This makes copter and plane both report Bad AHRS on boot in MP. They are now consistent so it would allow for the GCS to suppress the message.

Previously the ahrs could be not healthy and not initialised yet report healthy.

Due to plane having DCM plane the AHRS does not become healthy exactly as it its initialized as with copter, the DCM is always healthy thus resulting in the status bit going unhealthy when switched to ekf only once the ekf is initialized does it become healthy again.

This is probability a regression for copter, but i think the logic is better and will give a chance for GCS to fix. Alternately we could lie about the AHRS in a way that would work on all vehicles. such as:

```
if (ahrs.initialised()) {
    if (ahrs.healthy() && (!ahrs.have_inertial_nav() || ins.accel_calibrated_ok_all())) {
        control_sensors_health |= MAV_SYS_STATUS_AHRS;
    }
} else {
    control_sensors_health |= MAV_SYS_STATUS_AHRS;
}
```

IamPete1 commented:
follow up to https://github.com/ArduPilot/ardupilot/pull/13259

rmackay9 commented:
@IamPete1, so does this mean that Copter users can look forward to seeing "Bad AHRS" all the time at startup?  I really hope not because I'll get huge numbers of complaints.

IamPete1 commented:
@rmackay9 Yeah it does with current MP, but with a fix to MP we could suppress the message on boot for all vehicles. This change is, i think, correct to the definition of the mavlink message. ie how can we report a sensor healthy if its not initialized?

I'm abit on the fence, not sending a accurate status to fix a mission planner message for copter users 
seems a little backwards. That should really be done in MP, this enables that by making plane and copter constant. However we could also trivially fix the issue for all vehicles by a small logic change at this end at the expense of not sending accurate info.

rmackay9 commented:
@IamPete1, ok, from a practical point of view though, I'm definitely not happy with this going in until we figure out a way so "Bad AHRS" doesn't appear on MP's HUD.  Maybe @meee1 needs to be involved, I'm not sure but it's a scary message and I've already got enough support on my hands with Copter-4.0 that I can't handle anymore.

peterbarker commented:
@IamPete1 the logic you're changing here was especially crafted to *avoid* user support questions about the AHRS being red initially.  I don't like the current logic, but the reporting it bad initially was causing user support questions.

The thing to do here is probably to get the GCS authors to tone down their output while the vehicle is in initialising states, as reported by (e.g.) https://github.com/ardupilot/ardupilot/blob/master/APMrover2/GCS_Mavlink.cpp#L67

WickedShell commented:
I haven't flashed it to confirm it, but from a plane user point of view this probably doesn't represent a huge change. (Currently you start with no warning about AHRS, but as soon as you get GPS fix you get 10 seconds of bad AHRS warnings from the EKF).

IamPete1 commented:
Yeah, on plane you now get the warning from the start but it times out when it would have before.

IamPete1 commented:
We could also stay in the initializing mode for longer, although we would have to add it to copter.

davidbuzz commented:
Is good that the vehicles are consistent,  but the 'bad ahrs' message in the gcs needs to be changed to say 'initialising ahrs' or similar

IamPete1 commented:
possibly use (mavutil.mavlink.ESTIMATOR_CONST_POS_MODE and mavutil.mavlink.ESTIMATOR_ACCEL_ERROR) flags

rmackay9 commented:
Sadly we didn't come to a consensus on the dev call on how to fix this but my suggestion is that really the initial complaint is of the "Bad AHRS" message being displayed to Plane users of MP.  I think it might be resolved if MP was changed to somehow display an AHRS label next to EKF label on the HUD.  If this AHRS label turned red users probably wouldn't be so unhappy about it.

Pedals2Paddles commented:
IMO we should be providing accurate status to the GCS, and the GCS should accurately convey that status to the users.

- "BAD" means something is wrong and can't be used.
- Initializing is not something wrong. That's normal expected process, but also can't be used.
- Reporting bad when initializing is misinformation and a behavior change for copter.
- Reporting nothing when initializing is also not good, since it is denying the user information

We should simply report the AHRS status as initializing, good, or bad.  And have the GCS report that information as such.   IMO it's either that or leave it the way it is.

Naterater commented:
I am in full agreement with @Pedals2Paddles here... "BAD" needs to go away unless it is certain that the hardware is not functioning correctly. Missing GPS messages due to cold booting or a poor GPS environment is not indicative of "BAD" things on boot. That's normal of a system that is initializing.

IamPete1 commented:
rebased, the MP message is now 'Unhealthy AHRS' 

Plane master seems to no longer ever report  'Unhealthy AHRS'.

Copter master still does for a bit, but the message is not so scary as it was.

amilcarlucas commented:
Fails CI. Can you rebase this to re-test?

peterbarker commented:
.... so I wonder if we could just not show the AHRS as enabled (or even present?) while we're in initialising modes....

IamPete1 commented:
rebased and switched to only show AHRS as present once initialised as sugested by @peterbarker, this is slightly more honest that the current solution and still works to suppress the 'Unhealthy AHRS' messages. There is no change as far a the MP user can tell (unless there looking at the status flags in ctrl + f)

peterbarker commented:
Does present-but-not-enabled also work for surpressing the messages?  That may be better still.

IamPete1 commented:
I have swapped to present-but-not-enabled, both methods work to suppress the warning
